### PR TITLE
Savanna IBC tests improvements

### PR DESCRIPTION
**Pull request by systemzax**
Includes :

1) testing of action proofs
2) testing of finalizer policy change proofs, as well as testing of finality proofs with the newly changed finalizer policy generation

Addresses the following issues : 

https://github.com/AntelopeIO/spring/issues/119
https://github.com/AntelopeIO/spring/issues/134

### CODE REVIEW

File: libraries/chain/block_header_state.cpp [hunk 1]
```diff
@@ -79,6 +77,30 @@ const finalizer_policy& block_header_state::get_last_proposed_finalizer_policy()
...
+      };
+      for (auto ritr = finalizer_policies.rbegin(); ritr != finalizer_policies.rend(); ++ritr) {
+         if (ritr->second.state == finalizer_policy_tracker::state_t::pending) {
+            assert(highest_pending_generation() == ritr->second.policy);
+            return *ritr->second.policy;
+         }
+      }
+   }
+   return *active_finalizer_policy;
+}
```
Comment (greg7mdp) [inline libraries/chain/block_header_state.cpp:102]: Strange that these (Kevin's changes) show up in your PR. Probably need to pull main to your branch.

Comment (systemzax) [reply libraries/chain/block_header_state.cpp:102]: Yes, this PR depends on Kevin's PR, which I merged in here. I'll make sure to keep it up to date until it is merged.

Comment (greg7mdp) [reply libraries/chain/block_header_state.cpp:102]: Oh, my bad, for some reason I thought Kevin's PR was already merged. That explains it!

**Review by linh2931 — APPROVED** [review]

File: unittests/svnn_ibc_tests.cpp [hunk 2]
```diff
@@ -75,66 +75,217 @@ BOOST_AUTO_TEST_SUITE(svnn_ibc)
...
+      std::vector<digest_type> get_finality_leaves(const size_t cutoff){
+         assert(cutoff>=0 && cutoff<finality_leaves.size());
+         return std::vector<digest_type>(finality_leaves.begin(), finality_leaves.begin() + cutoff + 1);
+      }
+
+      // produce and propagate a block, update internal state as needed, and returns relevant IBC data 
+      ibc_block_data_t produce_block(){
+         auto result = this->produce_and_push_block_ex();
+         signed_block_ptr block = result.block;
+         action_trace onblock_trace = result.onblock_trace->action_traces[0];
```
Comment (heifner) [inline unittests/svnn_ibc_tests.cpp:173]: It is possible that the `onblock_trace` is nullptr. Should add a BOOST_REQUIRE that it exists.

File: unittests/svnn_ibc_tests.cpp [hunk 3]
```diff
@@ -75,66 +75,217 @@ BOOST_AUTO_TEST_SUITE(svnn_ibc)
...
+
+         // if we have policy diffs, process them
+         if (has_finalizer_policy_diffs(block)){
+
+            if (is_genesis){
+            // if block is genesis, the initial policy is the last proposed, last pending and currently active
```
Comment (heifner) [inline unittests/svnn_ibc_tests.cpp:179]: Indentation

**Review by heifner — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
group: IF
category: TEST
summary: Testing for IBC action proofs, and finalizer policy change proofs
Note:end

**Review by heifner — APPROVED** [review]

**Pull request merged**